instrumentation/net/http/otelhttpconv/DESIGN.md

```diff
@@ -0,0 +1,283 @@
+	//
+	// This method does not create a new span. It retrieves the current one from
+	// the context.
+	// It remains the instrumentation's responsibility to start and end spans.
+	Record(ctx context.Context, req *http.Request, w ResponseWrapper, cfg ...ServerRecordOption)
+}
+
+// ServerRecordSpanOption applies options to the RecordSpan method
+type ServerRecordOption interface{}
+```
+
+The `ClientRecordOption` and `ServerRecordOption` functional options allows
+passing optional parameters to be used within the `Record` method, such as the
+HTTP route.
+
+When the data those options provide is not specified, the related span
+attributes will not be set.
+
+### Compatibility
+
+These interfaces are kept relatively simple on purpose, so they can be used as
+foundations if folks wish to extend them with their own actual implementations.
+
+Once stable, we will not be able to change those interfaces anymore (adding,
+renaming or deleting methods), as that would break backwards compatibility with
+implementations.
+
+However, because they remain lightweight, they allow folks to do custom logic,
+such as:
+
+* A proxy that would allow combining multiple implementations together.
+* An implementation that covers unstable semantic conventions.
+* Implementations for older semantic conventions.
+* Implementations that cover custom implementations if a customer doesn't wish to use semconvs.
+
+### Implementations
+
+We will provide one official implementation of the described interfaces.
+This implementation will have the following requirements:
+
+* Support for the latest semantic conventions only.
+* Support for stable semantic conventions only.
+
+We may provide additional implementations later on such as:
+
+* An implementation that serves as a proxy to allow combining multiple implementations together.
+* An implementation that covers unstable semantic conventions.
+
+#### Example implementation
+
+The implementation example here is kept simple for the purpose of
+understandability.
+
+The following code sample provides a simple `http.Handler` that implements the
+provided interface to instrument HTTP applications.
+
+Because both the client and server interface are very similar, a client
+implementation would be similar too.
+
+```golang
+type middleware struct {
+	operation string
+
+	tracer      trace.Tracer
+	propagators propagation.TextMapPropagator
+	meter       metric.Meter
+	httpconv    otelhttpconv.Server
+}
+
+// NewMiddleware returns a tracing and metrics instrumentation middleware.
+// The handler returned by the middleware wraps a handler
+// in a span named after the operation and enriches it with metrics.
+func NewMiddleware(operation string) func(http.Handler) http.Handler {
+	m := middleware{
+		operation:   operation,
+		tracer:      otel.Tracer("http"),
+		propagators: otel.GetTextMapPropagator(),
+		meter:       otel.Meter("http"),
+		httpconv:    otelhttpconv.NewHTTPConv(otel.Tracer("httpconv"), otel.Meter("httpconv")),
```

Comment:
`NewHTTPConv` is not defined in the design.

Reply:
Sorry, that seemed clear to me. It creates an instance of our official implementation of the interface.
Since the implementation itself isn't in the design, I didn't add this method.